### ISSUE TITLE
App crashes upon restart if the app process has died in the background while on Case List or Case Feed screens

### ISSUE THREAD

**Issue by ogenez-tu** [forcedotcom/ServiceSDK-Android#97]
### Issue description
App crashes upon restart if the app process has died in the background while on Case List or Case Feed screens (CaseListActivity or CaseFeedActivity).

### Reproduction steps
* Build your app with some runtime permissions; let's use Location for this example
* Grant the Location permission to your app
* Open the app
* Authenticate into Salesforce
* Go to the Case List screen
* Leave the app via the Home button
* Go to the app settings and remove the "Location" permission; this will cause the app process to restart when coming back to the app
* Go back to the app, for example via the Recents screen

#### Expected results
The app's case list screen should be displayed in the same state it was in when it was left.

#### Actual results
The app crashes with the following stacktrace:

```
2018-10-30 09:57:44.231 13260-13260/<my app ID> E/AndroidRuntime: FATAL EXCEPTION: main
    Process: <my app ID>, PID: 13260
    java.lang.RuntimeException: Unable to start activity ComponentInfo{<my app ID>/com.salesforce.android.cases.ui.internal.features.caselist.CaseListActivity}: java.lang.IllegalStateException: You must configure with a CaseUIConfiguration prior to retrieving a CaseUIClient.
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.IllegalStateException: You must configure with a CaseUIConfiguration prior to retrieving a CaseUIClient.
        at com.salesforce.android.cases.ui.CaseUI.uiClient(CaseUI.java:96)
        at com.salesforce.android.cases.ui.internal.features.shared.BaseActivityDelegate.onCreate(BaseActivityDelegate.java:59)
        at com.salesforce.android.cases.ui.internal.features.caselist.CaseListActivity.onCreate(CaseListActivity.java:60)
        at android.app.Activity.performCreate(Activity.java:7009)
        at android.app.Activity.performCreate(Activity.java:7000)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
```

#### Additional info:
The same issue occurs when following the same steps and leaving the app while on a specific case feed instead. The stacktrace is mostly the same except that the activity failing to start is `CaseFeedActivity`:

```
2018-10-30 10:01:26.060 14882-14882/<my app ID> E/AndroidRuntime: FATAL EXCEPTION: main
    Process: <my app ID>, PID: 14882
    java.lang.RuntimeException: Unable to start activity ComponentInfo{<my app ID>/com.salesforce.android.cases.ui.internal.features.casefeed.CaseFeedActivity}: java.lang.IllegalStateException: You must configure with a CaseUIConfiguration prior to retrieving a CaseUIClient.
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.IllegalStateException: You must configure with a CaseUIConfiguration prior to retrieving a CaseUIClient.
        at com.salesforce.android.cases.ui.CaseUI.uiClient(CaseUI.java:96)
        at com.salesforce.android.cases.ui.internal.features.shared.BaseActivityDelegate.onCreate(BaseActivityDelegate.java:59)
        at com.salesforce.android.cases.ui.internal.features.casefeed.CaseFeedActivity.onCreate(CaseFeedActivity.java:65)
        at android.app.Activity.performCreate(Activity.java:7009)
        at android.app.Activity.performCreate(Activity.java:7000)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

### Version of SDK components being used
October 2018 Major Release of the Snap-Ins (Case Management version 4.1.2). We have observed it on previous versions of the SDK though.

### Device(s) and version(s) of Android being used
We have observed the issue on a variety of device, including Pixel/Android 8.0 and Nexus 6P/Android 8.1.

**Comment by rileymacdonald-sf** [comment]
@ogenez-tu thanks for providing such excellent details regarding this issue. Let me take some time to investigate and I'll get back to you with more information.

**Comment by evanisnor** [comment]
We've released version 4.1.3 of the Case SDK to solve the crash issue. If you repeat the reproduction steps after upgrading, you may notice that the last used Case Activity will appear blank. If this occurs, please consider initializing a `CaseUIClient` during `Application.onCreate()`.

Release notes:
https://github.com/forcedotcom/ServiceSDK-Android/releases/tag/216.0.1

Cheers,
Evan
